File: progpraxis_euler1.rb
# http://programmingpraxis.com/2015/02/10/project-euler-problem-1/

# iterates with enumerable
def euler_one_one(n,counter=0)
  (1...n).each { |number| counter += number if (number % 3 == 0 || number % 5 == 0) }
  counter
end

# recursive one liner
def euler_one_two(n,c=0,i=0)
  n == 1 ? c : (n % 3 == 0 || n % 5 == 0) && i != 0 ? euler_one_two(n-1,c+n,i+1) : euler_one_two(n-1,c,i+1)
end

# nested methods, map array
def euler_one_three(n)
  def three(x)
    x % 3 == 0
  end
  def five(x)
    x % 5 == 0
  end
  (1...n).collect { |i| i if (three(i) || five(i)) }.compact.inject(:+)
end

p euler_one_one(1000)
p euler_one_two(1000)
p euler_one_three(1000)

File: progpraxis_euler1.c
// http://programmingpraxis.com/2015/02/10/project-euler-problem-1/

#include <stdio.h>

// standard for loop
int eulerOneOne(int n)
{
    int accumulator = 0;
    for (int i = 1; i < n; i++) {
        if (i % 3 == 0 || i % 5 == 0) {
            accumulator += i;
        }
    }
    return accumulator;
}

// recursive
int eulerOneTwo(int n, int accumulator, int counter)
{
    if (n == 1) {
        return accumulator;
    } else if ((n % 3 == 0 || n % 5 == 0) && counter != 0) {
        return eulerOneTwo(n-1, accumulator+n, counter+1);
    } else {
        return eulerOneTwo(n-1, accumulator, counter+1);
    }
}

// generate and sum multiples of three and five
int eulerOneThree(int n)
{
    int accumulator = 0;
    for (int i = 0; i < n; i += 3) {
        accumulator += i;
    }
    for (int i = 0; i < n; i += 5) {
        if (i % 3 != 0) {
            accumulator += i;
        }
    }
    return accumulator;
}



int main(int argc, const char * argv[]) {
    int resultOne = eulerOneOne(1000);
    int resultTwo = eulerOneTwo(1000,0,0);
    int resultThree = eulerOneThree(1000);
    printf("%d, %d, %d\n", resultOne, resultTwo, resultThree);
    return 0;
}
